test/certs/mkcert.sh

```diff
@@ -288,4 +288,56 @@ gennocn() {
 	cert "$cert" "" -signkey "${key}.pem" -set_serial 1 -days -1 "$@"
 }
 
+genct() {
+    local OPTIND=1
+    local purpose=serverAuth
+
+    while getopts p: o
+    do
+        case $o in
+        p) purpose="$OPTARG";;
+        *) echo "Usage: $0 genee [-p EKU] cn keyname certname cakeyname cacertname" >&2
+           return 1;;
+        esac
+    done
+
+    shift $((OPTIND - 1))
+    local cn=$1; shift
+    local key=$1; shift
+    local cert=$1; shift
+    local cakey=$1; shift
+    local ca=$1; shift
+    local logkey=$1; shift
+
+    exts=$(printf "%s\n%s\n%s\n%s\n%s\n%s\n[alts]\n%s\n" \
+	    "subjectKeyIdentifier = hash" \
+	    "authorityKeyIdentifier = keyid, issuer" \
+	    "basicConstraints = CA:false" \
+	    "extendedKeyUsage = $purpose" \
+            "1.3.6.1.4.1.11129.2.4.3 = critical,ASN1:NULL"\
+	    "subjectAltName = @alts" "DNS=${cn}")
+    csr=$(req "$key" "CN = $cn") || return 1
+    echo "$csr" |
+	cert "$cert" "$exts" -CA "${ca}.pem" -CAkey "${cakey}.pem" \
+	    -set_serial 2 -days "${DAYS}" "$@"
+    cat ${cert}.pem ${ca}.pem > ${cert}-chain.pem
+    go run github.com/google/certificate-transparency-go/ctutil/sctgen \
+       --log_private_key ${logkey}.pem \
+       --timestamp="2020-01-01T00:00:00Z" \
+       --cert_chain ${cert}-chain.pem \
+       --tls_out ${cert}.tlssct
```

Comment:
This introduces a requirement on a live Internet connection, and running untrusted code from github.com into mkcert.sh. (And of course having "go" installed). While we tend to not run it during builds, and instead just cache the artefacts, I am not fond of this change. Perhaps just make it in a private branch, and commit just the output? Otherwise, we likely need actual OpenSSL code  that can generate this sort of object...

Reply:
As far as I understand, you don't actually need to have a live internet connection, you always need to have it checked out locally.

I do wish that we didn't need this external code to generate such a certificate, and it's probably not that hard to write something that does the same, but we currently don't have it.

Anyway, the reason to put it in that file is so that we at least have a documented way of creating such files.

Reply:
I agree with Victor.  Adding "go" as a dependency to build openssl seems bad.  Keep the lines commented output and just checkin the output.

Reply:
@richsalz Just to be clear, mkcert.sh is not normally run during builds. It is used to manufacture certificates for the test suite, which are then checked into git. So it is not actually needed for the vast majority of users...

Reply:
Er, we don't run this script when building, do we? This is a script to generate certs, which are checked in, isn't it? So kinda like "make update", but run even more seldom...

Reply:
thanks for the correction.  Is the project team okay with adding a go dependency?

Reply:
I am not fond of the `go` dependency, and would prefer to see it "go" away. But for that we'd need our own CT generation code (which probably should be added at some point). In the mean time this is only needed to generate test certs with SCTs. The rest of the cert generation bypasses this code.
So the real question is whether the `go` call should committed in tree, or just used privately, until we have our own code to do this.
My instinct was to not add the `go` call, but it is a very minor issue. This could e.g. just be commented out, as Kurt said, it is there primarily so that someone else (or he later) could retrace his steps.